File: examples/.net/core/shared/app_mcp_contracts.cs
namespace AppProtocol.Example.DotNet.Core.Shared;

public sealed record AppMcpClientInfo(string Name, string Version);

public sealed record AppMcpServerInfo(
    string Name,
    string Version,
    string ProtocolVersion,
    string? Instructions = null);

public sealed class AppMcpInitializeParams
{
    public required string ProtocolVersion { get; init; }
    public IDictionary<string, object?>? Capabilities { get; init; }
    public AppMcpClientInfo? ClientInfo { get; init; }
}

public sealed class AppMcpInitializeResult
{
    public required string ProtocolVersion { get; init; }
    public required IDictionary<string, object?> Capabilities { get; init; }
    public required IDictionary<string, object?> ServerInfo { get; init; }
    public string? Instructions { get; init; }
}

public sealed record AppMcpTextContent(string Type, string Text);

public sealed class AppMcpToolDescriptor
{
    public required string Name { get; init; }
    public string? Title { get; init; }
    public string? Description { get; init; }
    public required AppSchema InputSchema { get; init; }
    public AppSchema? OutputSchema { get; init; }
    public IDictionary<string, object?>? Annotations { get; init; }
}

public sealed class AppMcpResourceDescriptor
{
    public required string Uri { get; init; }
    public required string Name { get; init; }
    public string? Title { get; init; }
    public string? Description { get; init; }
    public string? MimeType { get; init; }
    public IDictionary<string, object?>? Annotations { get; init; }
}

public sealed class AppMcpTextResourceContent
{
    public required string Uri { get; init; }
    public string? MimeType { get; init; }
    public required string Text { get; init; }
}

public sealed class AppMcpCallResult
{
    public required IReadOnlyList<AppMcpTextContent> Content { get; init; }
    public object? StructuredContent { get; init; }
    public bool? IsError { get; init; }
}

public sealed class AppMcpReadResourceResult
{
    public required IReadOnlyList<AppMcpTextResourceContent> Contents { get; init; }
}

public sealed class AppMcpRequestContext
{
    public required string Transport { get; init; }
    public object? RequestId { get; init; }
    public string? SessionId { get; init; }
    public string? CorrelationId { get; init; }
    public AppMcpClientInfo? ClientInfo { get; init; }
    public string? ProtocolVersion { get; init; }
}

public interface IAppMcpServer
{
    AppMcpServerInfo ServerInfo();
    Task<AppMcpInitializeResult> InitializeAsync(AppMcpInitializeParams? parameters = null, AppMcpRequestContext? parent = null);
    Task<IReadOnlyList<AppMcpToolDescriptor>> ListToolsAsync(AppMcpRequestContext? parent = null);
    Task<IReadOnlyList<AppMcpResourceDescriptor>> ListResourcesAsync(AppMcpRequestContext? parent = null);
    Task<AppMcpReadResourceResult> ReadResourceAsync(string uri, AppMcpRequestContext? parent = null);
    Task<AppMcpCallResult> CallToolAsync(string name, object? arguments = null, AppMcpRequestContext? parent = null);
}

public abstract class BaseAppMcpAdapter
{
    public abstract string Transport { get; }
}

public abstract class BaseAppMcpProcessAdapter : BaseAppMcpAdapter
{
    public abstract Task ServeAsync(IAppMcpServer server, CancellationToken cancellationToken = default);
}

public sealed class AppMcpHttpExchange
{
    public required string Method { get; init; }
    public required string Path { get; init; }
    public required IDictionary<string, string?> Headers { get; init; }
    public string? BodyText { get; init; }
}

public sealed class AppMcpHttpResponse
{
    public required int StatusCode { get; init; }
    public IDictionary<string, string>? Headers { get; init; }
    public string? BodyText { get; init; }
}

public abstract class BaseAppMcpHttpAdapter : BaseAppMcpAdapter
{
    public abstract string EndpointPath { get; }

    public abstract Task<AppMcpHttpResponse?> HandleAsync(
        AppMcpHttpExchange exchange,
        IAppMcpServer server,
        CancellationToken cancellationToken = default);
}

public sealed class AppMcpProtocolError : Exception
{
    public AppMcpProtocolError(int code, string message, object? payload = null)
        : base(message)
    {
        Code = code;
        Payload = payload;
    }

    public int Code { get; }
    public object? Payload { get; }
}

File: src/core/shared/app_mcp_contracts.ts
/* ========================================================================== *
 * APP v1.1.2
 * core/shared/app_mcp_contracts.ts
 * ----------------------------------------------------------------------------
 * Contratos protocolares MCP para hosts agentic do APP.
 *
 * Estes contratos definem a fronteira estrutural entre o runtime do host
 * agentic e um ou mais adapters MCP concretos. Eles permanecem deliberadamente
 * agnósticos ao transporte:
 * - o host controla catálogo, policy e execução canônica
 * - o provider controla framing, lifecycle e mecânica do transporte
 *
 * Implementações concretas de MCP pertencem a registry._providers.
 * ========================================================================== */

import type { AppSchema, Dict } from "../domain.case";

export interface AppMcpClientInfo {
  name: string;
  version: string;
}

export interface AppMcpServerInfo {
  name: string;
  version: string;
  protocolVersion: string;
  instructions?: string;
}

export interface AppMcpInitializeParams {
  protocolVersion: string;
  capabilities?: Dict;
  clientInfo?: AppMcpClientInfo;
}

export interface AppMcpInitializeResult {
  protocolVersion: string;
  capabilities: {
    tools?: {
      listChanged?: boolean;
    };
    resources?: {
      listChanged?: boolean;
    };
  };
  serverInfo: {
    name: string;
    version: string;
  };
  instructions?: string;
}

export interface AppMcpTextContent {
  type: "text";
  text: string;
}

export interface AppMcpToolDescriptor {
  name: string;
  title?: string;
  description?: string;
  inputSchema: AppSchema;
  outputSchema?: AppSchema;
  annotations?: Dict;
}

export interface AppMcpResourceDescriptor {
  uri: string;
  name: string;
  title?: string;
  description?: string;
  mimeType?: string;
  annotations?: Dict;
}

export interface AppMcpTextResourceContent {
  uri: string;
  mimeType?: string;
  text: string;
}

export interface AppMcpCallResult {
  content: AppMcpTextContent[];
  structuredContent?: unknown;
  isError?: boolean;
}

export interface AppMcpReadResourceResult {
  contents: AppMcpTextResourceContent[];
}

export interface AppMcpRequestContext {
  transport: string;
  requestId?: string | number | null;
  sessionId?: string;
  correlationId?: string;
  clientInfo?: AppMcpClientInfo;
  protocolVersion?: string;
}

export interface AppMcpServer {
  serverInfo(): AppMcpServerInfo;
  initialize(
    params?: AppMcpInitializeParams,
    parent?: Partial<AppMcpRequestContext>
  ): Promise<AppMcpInitializeResult>;
  listTools(
    parent?: Partial<AppMcpRequestContext>
  ): Promise<AppMcpToolDescriptor[]>;
  listResources(
    parent?: Partial<AppMcpRequestContext>
  ): Promise<AppMcpResourceDescriptor[]>;
  readResource(
    uri: string,
    parent?: Partial<AppMcpRequestContext>
  ): Promise<AppMcpReadResourceResult>;
  callTool(
    name: string,
    args: unknown,
    parent?: Partial<AppMcpRequestContext>
  ): Promise<AppMcpCallResult>;
}

export abstract class BaseAppMcpAdapter {
  public abstract readonly transport: string;
}

export abstract class BaseAppMcpProcessAdapter extends BaseAppMcpAdapter {
  public abstract serve(server: AppMcpServer): Promise<void>;
}

export interface AppMcpHttpExchange {
  method: string;
  path: string;
  headers: Record<string, string | undefined>;
  bodyText?: string;
}

export interface AppMcpHttpResponse {
  statusCode: number;
  headers?: Record<string, string>;
  bodyText?: string;
}

export abstract class BaseAppMcpHttpAdapter extends BaseAppMcpAdapter {
  public abstract readonly endpointPath: string;
  public abstract handle(
    exchange: AppMcpHttpExchange,
    server: AppMcpServer
  ): Promise<AppMcpHttpResponse | undefined>;
}

export class AppMcpProtocolError extends Error {
  public readonly code: number;
  public readonly data?: unknown;

  constructor(code: number, message: string, data?: unknown) {
    super(message);
    this.name = "AppMcpProtocolError";
    this.code = code;
    this.data = data;
  }
}
